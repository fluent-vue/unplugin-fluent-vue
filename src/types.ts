interface ExternalPluginOptionsBase {
  locales: string[]
  checkSyntax?: boolean
}

interface ExternalPluginOptionsFolder extends ExternalPluginOptionsBase {
  baseDir: string
  ftlDir: string
}

interface ExternalPluginOptionsFunction extends ExternalPluginOptionsBase {
  getFtlPath: (locale: string, vuePath: string) => string
}

export type ExternalPluginOptions = ExternalPluginOptionsFolder | ExternalPluginOptionsFunction

export interface SFCPluginOptions {
  blockType?: string
  checkSyntax?: boolean
}

export interface InsertInfo {
  insertPos: number
  target: string
}
