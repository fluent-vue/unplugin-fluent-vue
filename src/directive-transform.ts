import type { DirectiveTransform } from '@vue/compiler-core'
import { createCompoundExpression, createSimpleExpression, NodeTypes } from '@vue/compiler-core'

export const directiveTransform: DirectiveTransform = (dir, node, context) => {
  const baseCode = `${context.prefixIdentifiers ? '_ctx.' : ''}`

  if (!dir.arg) {
    console.error(`[fluent-vue] Missing key argument for v-t directive`)
    return {
      props: [],
      needRuntime: true,
    }
  }

  const paramsExpr = dir.exp
    ? createCompoundExpression([dir.arg, createSimpleExpression(`, `, false), dir.exp])
    : dir.arg

  node.children = [
    {
      type: NodeTypes.INTERPOLATION,
      content: createCompoundExpression([
        createSimpleExpression(`${baseCode}$t(`, false),
        paramsExpr,
        createSimpleExpression(`)`, false),
      ]),
      loc: node.loc,
    },
  ]

  return {
    props: [],
    needRuntime: true,
  }
}
