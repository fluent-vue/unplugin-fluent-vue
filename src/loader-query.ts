export interface VueQuery {
  vue?: boolean
  type?: 'script' | 'template' | 'style' | 'custom' | 'fluent'
  blockType?: string
  index?: number
  locale?: string
}

export function parseVueRequest(id: string) {
  const [filename, rawQuery] = id.split('?', 2)
  const params = new URLSearchParams(rawQuery)
  const ret: VueQuery = {}

  ret.vue = params.has('vue')

  if (params.has('type'))
    ret.type = params.get('type') as VueQuery['type']

  if (params.has('blockType'))
    ret.blockType = params.get('blockType')

  if (params.has('index'))
    ret.index = Number(params.get('index'))

  if (params.has('locale'))
    ret.locale = params.get('locale')

  return {
    filename,
    query: ret,
  }
}

export function isCustomBlock(query: VueQuery, options: { blockType: string }): boolean {
  return (
    'vue' in query
    && (query.type === 'custom' // for vite (@vite-plugin-vue)
    || query.type === options.blockType // for webpack (vue-loader)
    || query.blockType === options.blockType) // for webpack (vue-loader)
  )
}
