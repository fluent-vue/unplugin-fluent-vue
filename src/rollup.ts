export { rollupPlugin as ExternalFluentPlugin } from './plugins/external-plugin'
export { rollupPlugin as SFCFluentPlugin } from './plugins/sfc-plugin'

export type { ExternalPluginOptions, SFCPluginOptions } from './types'
