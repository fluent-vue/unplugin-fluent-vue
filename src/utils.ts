import { promises as fs } from 'node:fs'

export function warn(...args: unknown[]) {
  console.warn('[unplugin-fluent-vue] ', ...args)
}

export function error(...args: unknown[]) {
  console.error('[unplugin-fluent-vue] ', ...args)
}

export async function getRaw(path: string): Promise<string> {
  return fs.readFile(path, { encoding: 'utf-8' })
}

export function raiseError(message: string) {
  throw new Error(`[unplugin-fluent-vue] ${message}`)
}
