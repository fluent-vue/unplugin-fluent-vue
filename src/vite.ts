export { vitePlugin as ExternalFluentPlugin } from './plugins/external-plugin'
export { vitePlugin as SFCFluentPlugin } from './plugins/sfc-plugin'

export type { ExternalPluginOptions, SFCPluginOptions } from './types'
