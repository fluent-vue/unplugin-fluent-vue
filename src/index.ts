import { join, relative } from 'path'
import { stat as fsStat } from 'fs/promises'
import type { UnpluginContextMeta, VitePlugin } from 'unplugin'
import { createUnplugin } from 'unplugin'

import { parse } from '@vue/compiler-sfc'

import MagicString from 'magic-string'
import { createFilter, makeLegalIdentifier } from '@rollup/pluginutils'

import createDebug from 'debug'

import { getRaw, raiseError } from './utils'

const debug = createDebug('unplugin-fluent-vue')

export interface UserOptions {
  blockType?: string
  external?: {
    baseDir: string
    ftlDir: string
    locales: string[]
  }
}

export interface VueQuery {
  vue?: boolean
  src?: boolean
  global?: boolean
  type?: 'script' | 'template' | 'style' | 'custom' | 'fluent'
  blockType?: string
  index?: number
  locale?: string
  raw?: boolean
  issuerPath?: string
}

function isCustomBlock(query: VueQuery, options: UserOptions): boolean {
  return (
    'vue' in query
    && (query.type === 'custom' // for vite (@vite-plugin-vue)
      || query.type === options.blockType // for webpack (vue-loader)
      || query.blockType === options.blockType) // for webpack (vue-loader)
  )
}

interface InsertInfo {
  insertPos: number
  target: string
}

function getInsertInfo(source: string): InsertInfo {
  let target = null

  // vite-plugin-vue2
  if (source.includes('__component__'))
    target = '__component__'

  // rollup-plugin-vue
  if (source.includes('export default script'))
    target = 'script'

  // @vitejs/plugin-vue
  if (source.includes('_sfc_main'))
    target = '_sfc_main'

  const insertPos = source.indexOf('export default')

  if (insertPos === -1 || target === null)
    throw new Error('Could not parse vue component. This is the issue with unplugin-fluent-vue.\nPlease report this issue to the unplugin-fluent-vue repository.')

  return { insertPos, target }
}

export function parseVueRequest(id: string) {
  const [filename, rawQuery] = id.split('?', 2)
  const params = new URLSearchParams(rawQuery)
  const ret: VueQuery = {}

  ret.vue = params.has('vue')
  ret.global = params.has('global')
  ret.src = params.has('src')
  if (params.has('type'))
    ret.type = params.get('type') as VueQuery['type']

  if (params.has('blockType'))
    ret.blockType = params.get('blockType')

  if (params.has('index'))
    ret.index = Number(params.get('index'))

  if (params.has('locale'))
    ret.locale = params.get('locale')

  if (params.has('issuerPath'))
    ret.issuerPath = params.get('issuerPath')

  return {
    filename,
    query: ret,
  }
}

async function fileExists(filename: string): Promise<boolean> {
  try {
    const stat = await fsStat(filename, { throwIfNoEntry: false } as any)
    return !!stat
  }
  catch {
    return false
  }
}

async function getCode(
  source: string,
  filename: string,
  sourceMap: boolean,
  query: VueQuery,
  framework: UnpluginContextMeta['framework'] = 'vite',
): Promise<string> {
  const { index, issuerPath } = query
  if (!Number.isInteger(index))
    raiseError(`unexpected index: ${index}`)

  if (framework === 'webpack') {
    if (issuerPath) {
      // via `src=xxx` of SFC
      debug(`getCode (webpack) ${index} via issuerPath`, issuerPath)
      return await getRaw(filename)
    }
    else {
      const result = parse(await getRaw(filename), {
        sourceMap,
        filename,
      })
      const block = result.descriptor.customBlocks[index!]
      if (block) {
        const code = block.src ? await getRaw(block.src) : block.content
        debug(`getCode (webpack) ${index} from SFC`, code)
        return code
      }
      else {
        return source
      }
    }
  }
  else {
    return source
  }
}

function normalizePath(path: string) {
  return path.replace(/\\/g, '/')
}

const isVue = createFilter(['**/*.vue'])

export const unplugin = createUnplugin((options: UserOptions, meta) => {
  const resolvedOptions: Required<UserOptions> = {
    blockType: 'fluent',
    external: null,
    ...options,
  }

  return {
    name: 'unplugin-fluent-vue',
    enforce: 'post',
    transformInclude(id: string) {
      return isVue(id)
    },
    async transform(source: string, id: string) {
      const { filename, query } = parseVueRequest(id)

      if (isCustomBlock(query, resolvedOptions)) {
        const code = await getCode(source, filename, false, query, meta.framework)

        // vue-loader pads SFC file sections with newlines - trim those
        const data = code.replace(/^(\n|\r\n)+|(\n|\r\n)+$/g, '')

        const hotCode = meta.watchMode
          ? `
if (module.hot) {
  delete target._fluent
  if (typeof __VUE_HMR_RUNTIME__ !== 'undefined') {
    // Vue 3
    const id = target.__hmrId
    const api = __VUE_HMR_RUNTIME__
    api.reload(id, target)
  } else {
    // Vue 2
    // There is no proper api to access HMR for component from custom block
    // so use this magic
    delete target._Ctor
  }
}`
          : ''

        if (query.locale == null)
          this.error('Custom block does not have locale attribute')

        return `
import { FluentResource } from '@fluent/bundle'

export default function (Component) {
  const target = Component.options || Component
  target.fluent = target.fluent || {}
  target.fluent['${query.locale}'] = new FluentResource(\`${data}\`)
  ${hotCode}
}\n`
      }

      if (isVue(id) && resolvedOptions.external != null) {
        const external = resolvedOptions.external

        const relativePath = relative(external.baseDir, id)

        const magic = new MagicString(source, { filename: id })

        const existingTranslations = []
        for (const locale of external.locales) {
          const ftlPath = normalizePath(join(external.ftlDir, locale, `${relativePath}.ftl`))
          const ftlExists = await fileExists(ftlPath)

          if (ftlExists) {
            this.addWatchFile(ftlPath)

            existingTranslations.push(locale)
            magic.prepend(`import ${makeLegalIdentifier(locale)}_ftl from '${ftlPath}';\n`)
          }
        }

        const { insertPos, target } = getInsertInfo(source)

        magic.prepend('\n')
        magic.appendLeft(insertPos, `${target}.fluent = ${target}.fluent || {};\n`)
        for (const locale of existingTranslations)
          magic.appendLeft(insertPos, `${target}.fluent['${locale}'] = ${makeLegalIdentifier(locale)}_ftl\n`)

        return {
          code: magic.toString(),
          map: magic.generateMap({ hires: true }),
        }
      }

      return undefined
    },
  }
})

export const vitePlugin: VitePlugin = unplugin.vite
export const rollupPlugin = unplugin.rollup
export const webpackPlugin = unplugin.webpack
export const esbuildPlugin = unplugin.esbuild
