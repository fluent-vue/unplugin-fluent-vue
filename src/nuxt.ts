// Workaround for:
// src/nuxt.ts(5,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/@nuxt+schema@3.0.0_rollup@2.79.0/node_modules/@nuxt/schema'. This is likely not portable. A type annotation is necessary.
import type { } from '@nuxt/schema'
import type { ExternalPluginOptionsFolder, ExternalPluginOptionsFunction } from './types'

import { addVitePlugin, defineNuxtModule } from '@nuxt/kit'
import { directiveTransform } from './directive-transform'
import { ExternalFluentPlugin, SFCFluentPlugin, type SFCPluginOptions } from './vite'

interface NuxtFluentOptions {
  sfc?: SFCPluginOptions
  external?: Omit<ExternalPluginOptionsFolder, 'baseDir'> | ExternalPluginOptionsFunction
  /**
   * @default 't' v-t directive name for the directive transform
   */
  directiveName?: string
}

export default defineNuxtModule<NuxtFluentOptions>({
  meta: {
    name: 'fluent-vue',
    configKey: 'fluentVue',
  },
  setup(options, nuxt) {
    if (!options.sfc && !options.external) {
      console.error(`[fluent-vue/nuxt] You need to enable at least one of the fluent-vue plugins`)
      return
    }

    nuxt.options.vue.compilerOptions.directiveTransforms ??= {}
    nuxt.options.vue.compilerOptions.directiveTransforms[options.directiveName ?? 't'] = directiveTransform

    if (options.sfc)
      addVitePlugin(SFCFluentPlugin(options.sfc))

    if (options.external) {
      const externalOptions = {
        ...options.external,
        baseDir: nuxt.options.srcDir,
      }
      addVitePlugin(ExternalFluentPlugin(externalOptions))
    }
  },
})

declare module '@nuxt/schema' {
  interface NuxtConfig {
    fluentVue?: NuxtFluentOptions
  }
  interface NuxtOptions {
    fluentVue?: NuxtFluentOptions
  }
}
