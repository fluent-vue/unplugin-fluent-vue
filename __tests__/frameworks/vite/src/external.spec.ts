import { resolve } from 'path'
import { describe, expect, it } from 'vitest'

import vue3base from '@vitejs/plugin-vue'
import compiler from '@vue/compiler-sfc'

import { ExternalFluentPlugin } from '../../../../src/vite'
import { compile } from './util'

const vue3 = () => vue3base({
  compiler,
})

const baseDir = resolve(__dirname, '../../..')

describe('Vite external', () => {
  it('works', async () => {
    // Arrange
    // Act
    const code = await compile({
      plugins: [
        vue3(),
        ExternalFluentPlugin({
          baseDir: resolve(baseDir, 'fixtures'),
          ftlDir: resolve(baseDir, 'fixtures/ftl'),
          locales: ['en', 'da'],
        }),
      ],
    }, '/fixtures/components/external.vue')

    // Assert
    expect(code).toMatchSnapshot()
  })

  it('works with script setup', async () => {
    // Arrange
    // Act
    const code = await compile({
      plugins: [
        vue3(),
        ExternalFluentPlugin({
          baseDir: resolve(baseDir, 'fixtures'),
          ftlDir: resolve(baseDir, 'fixtures/ftl'),
          locales: ['en', 'da'],
        }),
      ],
    }, '/fixtures/components/external.setup.vue')

    // Assert
    expect(code).toMatchSnapshot()
  })
})
