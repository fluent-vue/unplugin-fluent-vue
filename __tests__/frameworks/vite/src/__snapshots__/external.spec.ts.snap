// Vitest Snapshot v1

exports[`Vite external > works 1`] = `
"import { createHotContext as __vite__createHotContext } from \\"/@vite/client\\";import.meta.hot = __vite__createHotContext(\\"/fixtures/components/external.vue\\");import da_ftl from '/fixtures/ftl/da/components/external.vue.ftl?import';
import en_ftl from '/fixtures/ftl/en/components/external.vue.ftl?import';
const _sfc_main = {}
import { toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, openBlock as _openBlock, createElementBlock as _createElementBlock } from \\"/@id/vue\\"

function _sfc_render(_ctx, _cache) {
  return (_openBlock(), _createElementBlock(\\"div\\", null, [
    _createElementVNode(\\"div\\", null, _toDisplayString(_ctx.$t('hello-user', { userName: _ctx.userName })), 1 /* TEXT */),
    _createElementVNode(\\"div\\", null, _toDisplayString(_ctx.$t('shared-photos', { userName: _ctx.userName, photoCount: _ctx.photoCount, userGender: _ctx.userGender })), 1 /* TEXT */)
  ]))
}


_sfc_main.__hmrId = \\"7d038643\\"
typeof __VUE_HMR_RUNTIME__ !== 'undefined' && __VUE_HMR_RUNTIME__.createRecord(_sfc_main.__hmrId, _sfc_main)
import.meta.hot.accept(mod => {
  if (!mod) return
  const { default: updated, _rerender_only } = mod
  if (_rerender_only) {
    __VUE_HMR_RUNTIME__.rerender(updated.__hmrId, updated.render)
  } else {
    __VUE_HMR_RUNTIME__.reload(updated.__hmrId, updated)
  }
})
import _export_sfc from '/@id/__x00__plugin-vue:export-helper'
_sfc_main.fluent = _sfc_main.fluent || {};
_sfc_main.fluent['en'] = en_ftl
_sfc_main.fluent['da'] = da_ftl

const __HOT_API__ = import.meta.hot || import.meta.webpackHot
if (__HOT_API__) {
  __HOT_API__.accept(['/fixtures/ftl/en/components/external.vue.ftl', '/fixtures/ftl/da/components/external.vue.ftl'], () => {
    _sfc_main.fluent['en'] = en_ftl
_sfc_main.fluent['da'] = da_ftl

    delete _sfc_main._fluent
    if (typeof __VUE_HMR_RUNTIME__ !== 'undefined') {
      // Vue 3
      __VUE_HMR_RUNTIME__.reload(_sfc_main.__hmrId, _sfc_main)
    } else {
      // Vue 2
      // There is no proper api to access HMR for component from custom block
      // so use this magic
      delete _sfc_main._Ctor
    }
  })
}
export default /*#__PURE__*/_export_sfc(_sfc_main, [['render',_sfc_render],['__file',\\"/fixtures/components/external.vue\\"]])"
`;

exports[`Vite external > works with script setup 1`] = `
"import { createHotContext as __vite__createHotContext } from \\"/@vite/client\\";import.meta.hot = __vite__createHotContext(\\"/fixtures/components/external.setup.vue\\");import en_ftl from '/fixtures/ftl/en/components/external.setup.vue.ftl?import';
const _sfc_main = {
  __name: 'external.setup',
  setup(__props, { expose }) {
  expose();

  const test = 'test'

const __returned__ = { test }
Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true })
return __returned__
}

}
import { toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, openBlock as _openBlock, createElementBlock as _createElementBlock } from \\"/@id/vue\\"

function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return (_openBlock(), _createElementBlock(\\"div\\", null, [
    _createElementVNode(\\"div\\", null, _toDisplayString(_ctx.$t('hello-user', { userName: _ctx.userName })), 1 /* TEXT */),
    _createElementVNode(\\"div\\", null, _toDisplayString(_ctx.$t('shared-photos', { userName: _ctx.userName, photoCount: _ctx.photoCount, userGender: _ctx.userGender })), 1 /* TEXT */)
  ]))
}


_sfc_main.__hmrId = \\"38704845\\"
typeof __VUE_HMR_RUNTIME__ !== 'undefined' && __VUE_HMR_RUNTIME__.createRecord(_sfc_main.__hmrId, _sfc_main)
import.meta.hot.accept(mod => {
  if (!mod) return
  const { default: updated, _rerender_only } = mod
  if (_rerender_only) {
    __VUE_HMR_RUNTIME__.rerender(updated.__hmrId, updated.render)
  } else {
    __VUE_HMR_RUNTIME__.reload(updated.__hmrId, updated)
  }
})
import _export_sfc from '/@id/__x00__plugin-vue:export-helper'
_sfc_main.fluent = _sfc_main.fluent || {};
_sfc_main.fluent['en'] = en_ftl

const __HOT_API__ = import.meta.hot || import.meta.webpackHot
if (__HOT_API__) {
  __HOT_API__.accept(['/fixtures/ftl/en/components/external.setup.vue.ftl'], () => {
    _sfc_main.fluent['en'] = en_ftl

    delete _sfc_main._fluent
    if (typeof __VUE_HMR_RUNTIME__ !== 'undefined') {
      // Vue 3
      __VUE_HMR_RUNTIME__.reload(_sfc_main.__hmrId, _sfc_main)
    } else {
      // Vue 2
      // There is no proper api to access HMR for component from custom block
      // so use this magic
      delete _sfc_main._Ctor
    }
  })
}
export default /*#__PURE__*/_export_sfc(_sfc_main, [['render',_sfc_render],['__file',\\"/fixtures/components/external.setup.vue\\"]])"
`;
